File: position.js
const _ = require('lodash');

const GetAdjacentTiles = ({ x, y }) => {
	return {
		up:    { x, y: y + 1 },
		down:  { x, y: y - 1 },
		left:  { x: x - 1, y },
		right: { x: x + 1, y },
	};
};

const IsOutsideBoard = ({ x, y }, board) => {
	return x < 0 || y < 0 || x >= board.width || y >= board.height;
};

const IsDeadly = (pos, { board, you }) => {
	if (IsOutsideBoard(pos, board)) {
		return true;
	}

	const selfCollision = _.some(_.initial(you.body), pos);
	if (selfCollision) {
		return true;
	}

	const anySnakeCollision = _.some(board.snakes, (snake) => {
		const collision = _.some(_.initial(snake.body), pos);
		if (!collision) {
			return false;
		}

		if (Matches(pos, snake.head) && you.length > snake.length) {
			return false;
		}

		return true;
	});
	if (anySnakeCollision) {
		return true;
	}

	return false;
};

const IsRisky = (pos, { board, you }) => {
	return _.some(board.snakes, (snake) => {
		const adjacent = GetAdjacentTiles(snake.head);
		const gettinSpicy = _.some(adjacent, pos);

		return gettinSpicy && you.length <= snake.length;
	});
};

const IsSafe = (pos, context) => {
	return !IsDeadly(pos, context) && !IsRisky(pos, context);
};

const Matches = (posA, posB) => {
	return _.isEqual(posA, posB);
};

const ToDirection = (to, from) => {
	const x = Math.sign(to.x - from.x);
	const y = Math.sign(to.y - from.y);

	if (x !== 0 && y !== 0) {
		return null;
	}

	if (x !== 0) {
		return x > 0 ? 'right' : 'left';
	}

	return y > 0 ? 'up' : 'down';
};

module.exports = {
	GetAdjacentTiles,
	IsOutsideBoard,
	IsDeadly,
	IsRisky,
	IsSafe,
	Matches,
	ToDirection,
};

File: _/position/position.go
package position

import (
	snek "github.com/JMTyler/battlesnake/_"
)

func GetAdjacentTiles(pos snek.Position) map[string]snek.Position {
	// HACK: Should replace these magic numbers with context.Board.Width somehow.
	cells := make(map[string]snek.Position)
	if pos.Y < 10 {
		cells["up"] = snek.Position{pos.X, pos.Y + 1}
	}
	if pos.Y > 0 {
		cells["down"] = snek.Position{pos.X, pos.Y - 1}
	}
	if pos.X > 0 {
		cells["left"] = snek.Position{pos.X - 1, pos.Y}
	}
	if pos.X < 10 {
		cells["right"] = snek.Position{pos.X + 1, pos.Y}
	}
	return cells
}

func IsOutsideBoard(pos snek.Position, board snek.Board) bool {
	return pos.X < 0 || pos.Y < 0 || pos.X >= board.Width || pos.Y >= board.Height
}

func IsDeadly(pos snek.Position, context snek.Context) bool {
	if IsOutsideBoard(pos, context.Board) {
		return true
	}

	for _, cell := range context.You.Body[1 : len(context.You.Body)-1] {
		if cell == pos {
			// self collision
			return true
		}
	}

	for _, snake := range context.Board.Snakes {
		collision := false
		for _, cell := range snake.Body[:len(snake.Body)-1] {
			if cell == pos {
				collision = true
			}
		}

		if collision {
			if pos != snake.Head || context.You.Length <= snake.Length {
				return true
			}
		}
	}

	return false
}

func IsRisky(pos snek.Position, context snek.Context) bool {
	for _, snake := range context.Board.Snakes {
		adjacent := GetAdjacentTiles(snake.Head)
		// TODO: should be able to use range to iterate over the adjacent map instead
		gettinSpicy := pos == adjacent["left"] ||
			pos == adjacent["right"] ||
			pos == adjacent["up"] ||
			pos == adjacent["down"]

		if gettinSpicy {
			if context.You.Length <= snake.Length {
				return true
			}
		}
	}

	return false
}

func IsSafe(pos snek.Position, context snek.Context) bool {
	return !IsDeadly(pos, context) && !IsRisky(pos, context)
}

func Matches(posA snek.Position, posB snek.Position) bool {
	return posA == posB
}

func ToDirection(to snek.Position, from snek.Position) string {
	x := to.X - from.X
	y := to.Y - from.Y

	if x != 0 && y != 0 {
		// TODO: error: non-lateral
		return ""
	}

	if x > 0 {
		return "right"
	}

	if x < 0 {
		return "left"
	}

	if y > 0 {
		return "up"
	}

	return "down"
}
